File: GameStoreAndroid/app/src/main/java/johnny/tutorial/gamestoreandroid/ProductListActivity.java
package johnny.tutorial.gamestoreandroid;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import johnny.tutorial.gamestoreandroid.constant.GameStoreConstants.ViewMode;
import johnny.tutorial.gamestoreandroid.db.ProductDbHelper;
import johnny.tutorial.gamestoreandroid.model.Product;

import java.util.ArrayList;

import johnny.tutorial.gamestoreandroid.constant.GameStoreConstants;

public class ProductListActivity extends AppCompatActivity {

    private static final String TAG = "ProductListActivity";
    private ProductDbHelper mHelper;
    private ListView mListView;
    private ListAdapter mAdapter;
    private ViewMode mMode;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_product_list);

        mMode = ViewMode.Display;
        mHelper = new ProductDbHelper(this);
        mListView = (ListView) findViewById(R.id.list_product);

        updateUI();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (mMode == ViewMode.Display) {
            getMenuInflater().inflate(R.menu.list_menu, menu);
            return super.onCreateOptionsMenu(menu);
        } else {
            getMenuInflater().inflate(R.menu.edit_menu, menu);
            return super.onCreateOptionsMenu(menu);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_addproduct:
                Log.d(TAG, "Add a new product");
                Intent intent = new Intent(this, ProductDetailActivity.class);
                intent.putExtra(GameStoreConstants.ParamAction, GameStoreConstants.ParamActionAdd);
                this.startActivity(intent);
                return true;
            case R.id.action_editproduct:
                Log.d(TAG, "Switch to Edit mode");
                mMode = ViewMode.Edit;
                updateUI();
                return true;
            case R.id.action_delete:
                Log.d(TAG, "Delete selected products");
                if (DeleteSelectedItems() == false) {
                    Toast.makeText(this, "Select at least one item to delete", Toast.LENGTH_LONG).show();
                    return false;
                }
                mMode = ViewMode.Display;
                updateUI();
                return true;
            case R.id.action_cancel:
                Log.d(TAG, "Cancel edit");
                mMode = ViewMode.Display;
                updateUI();
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    // Product Adapter
    private class ProductAdapter extends BaseAdapter {
        private Context context;
        private int resource;
        private ArrayList<Product> productList;
        private ViewMode mode;
        private LayoutInflater inflater;

        ProductAdapter(Context context, int resource, ArrayList<Product> productList, ViewMode mode) {
            this.context = context;
            this.resource = resource;
            this.productList = productList;
            this.mode = mode;
            this.inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        }

        @Override
        public int getCount() {
            return productList.size();
        }

        @Override
        public Object getItem(int i) {
            return productList.get(i);
        }

        @Override
        public long getItemId(int i) {
            return i;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            ViewHolder holder;
            View row = convertView;
            if (row == null) {
                row = inflater.inflate(this.resource, parent, false);
                holder = new ViewHolder();
                holder.productname = (TextView) row.findViewById(R.id.productname);
                holder.price = (TextView) row.findViewById(R.id.price);
                holder.image = (ImageView) row.findViewById(R.id.image);
                holder.checkbox = (CheckBox) row.findViewById(R.id.checkbox);
                row.setTag(holder);
            } else {
                holder = (ViewHolder) row.getTag();
            }

            Product product = productList.get(position);
            holder.id = product.getProductId();
            holder.productname.setText(product.getProductName());
            holder.price.setText("$" + String.valueOf(product.getPrice()));
            holder.image.setImageBitmap(product.getImage());
            if (mode == ViewMode.Display) {
                holder.checkbox.setVisibility(View.GONE);
            } else {
                holder.checkbox.setVisibility(View.VISIBLE);
            }
            row.setOnClickListener(new ProductItemOnClickListener(context, product.getProductId()));
            return row;
        }

        class ViewHolder {
            int id;
            TextView productname;
            TextView price;
            ImageView image;
            CheckBox checkbox;
        }

    }

    private class ProductItemOnClickListener implements View.OnClickListener {
        private Context context;
        private int id;

        public ProductItemOnClickListener(Context context, int id) {
            this.context = context;
            this.id = id;
        }

        @Override
        public void onClick(View view) {
            Intent intent = new Intent(this.context, ProductDetailActivity.class);
            intent.putExtra(GameStoreConstants.ParamAction, GameStoreConstants.ParamActionEdit);
            intent.putExtra(GameStoreConstants.ParamId, id);
            this.context.startActivity(intent);
        }
    }

    private void updateUI() {
        ArrayList<Product> productList = mHelper.getAllProducts();
        if (productList.size() == 0) {
            productList = createDummyData();
        }

        mAdapter = new ProductAdapter(this, R.layout.product_list_item, productList, mMode);
        mListView.setAdapter(mAdapter);
        invalidateOptionsMenu();
    }

    private Boolean DeleteSelectedItems() {
        boolean isSelected = false;
        for (int i = 0; i < mListView.getCount(); i++) {
            View rowView = mListView.getChildAt(i);
            CheckBox cb = (CheckBox) rowView.findViewById(R.id.checkbox);
            if (cb.isChecked()) {
                isSelected = true;
                ProductAdapter.ViewHolder holder = (ProductAdapter.ViewHolder) rowView.getTag();
                mHelper.deleteProduct(holder.id);
            }
        }
        return isSelected;
    }

    private ArrayList<Product> createDummyData() {
        Bitmap image1 = BitmapFactory.decodeResource(getResources(), R.drawable.xbox360);
        mHelper.insertProduct("xbox360", 299.00, image1);
        Bitmap image2 = BitmapFactory.decodeResource(getResources(), R.drawable.wii);
        mHelper.insertProduct("wii", 269.00, image2);
        Bitmap image3 = BitmapFactory.decodeResource(getResources(), R.drawable.controller);
        mHelper.insertProduct("controller", 19.99, image3);
        return mHelper.getAllProducts();
    }
}

File: GameStoreXamarin/GameStoreXamarin.Android/ProductListActivity.cs
using System;
using System.Collections.Generic;

using Android.App;
using Android.Content;
using Android.OS;
using Android.Views;
using Android.Widget;
using Android.Graphics;
using GameStoreXamarin.Core.Models;
using static GameStoreXamarin.Android.GameStoreConstants;
using static Android.Views.View;
using Android.Util;
using GameStoreXamarin.Android.Helper;
using GameStoreXamarin.Core.Data;

namespace GameStoreXamarin.Android
{
    [Activity(Label = "ProductListActivity", MainLauncher = true)]
    public class ProductListActivity : Activity
    {
        private const String TAG = "ProductListActivity";
        private ListView mListView;
        private BaseAdapter mAdapter;
        private ListViewMode mMode;
        private List<Product> _productList;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            Xamarin.Forms.Forms.Init(this, savedInstanceState);
            SetContentView(Resource.Layout.ProductList);

            mMode = ListViewMode.Display;
            mListView = (ListView)FindViewById(Resource.Id.list_product);

            updateUI();
        }

        public override Boolean OnCreateOptionsMenu(IMenu menu)
        {
            if (mMode == ListViewMode.Display)
            {
                MenuInflater.Inflate(Resource.Menu.list_menu, menu);
                return base.OnCreateOptionsMenu(menu);
            }
            else
            {
                MenuInflater.Inflate(Resource.Menu.edit_menu, menu);
                return base.OnCreateOptionsMenu(menu);
            }
        }

        public override Boolean OnOptionsItemSelected(IMenuItem item)
        {
            switch (item.ItemId)
            {
                case Resource.Id.action_addproduct:
                    Log.Debug(TAG, "Add a new product");
                    var intent = new Intent(this, typeof(ProductDetailActivity));
                    intent.PutExtra(GameStoreConstants.ParamAction, GameStoreConstants.ParamActionAdd);
                    this.StartActivity(intent);
                    return true;
                case Resource.Id.action_editproduct:
                    Log.Debug(TAG, "Switch to Edit mode");
                    mMode = ListViewMode.Edit;
                    updateUI();
                    return true;
                case Resource.Id.action_delete:
                    Log.Debug(TAG, "Delete selected products");
                    if (DeleteSelectedItems() == false) {
                        Toast.MakeText(this, "Select at least one item to delete", ToastLength.Long).Show();
                        return false;
                    }
                    mMode = ListViewMode.Display;
                    updateUI();
                    return true;
                case Resource.Id.action_cancel:
                    Log.Debug(TAG, "Cancel edit");
                    mMode = ListViewMode.Display;
                    updateUI();
                    return true;
                default:
                    return base.OnOptionsItemSelected(item);
            }
        }

        private void updateUI()
        {
            _productList = DatabaseHelper.Database.GetProducts();
            if (_productList.Count == 0)
            {
                CreateDummyData();
            }

            mAdapter = new ProductAdapter(this, Resource.Layout.ProductListItem, _productList, mMode);
            mListView.Adapter = mAdapter;
            InvalidateOptionsMenu();
        }

        private void CreateDummyData()
        {
            Product product = new Product();
            product.ProductName = "xbox360";
            product.Price = 299.00;
            product.Image = ImageHelper.BitmapToBytes(BitmapFactory.DecodeResource(Resources, Resource.Drawable.xbox360));
            DatabaseHelper.Database.SaveProduct(product);

            product = new Product();
            product.ProductName = "wii";
            product.Price = 269.00;
            product.Image = ImageHelper.BitmapToBytes(BitmapFactory.DecodeResource(Resources, Resource.Drawable.wii));
            DatabaseHelper.Database.SaveProduct(product);

            product = new Product();
            product.ProductName = "controller";
            product.Price = 19.99;
            product.Image = ImageHelper.BitmapToBytes(BitmapFactory.DecodeResource(Resources, Resource.Drawable.controller));
            DatabaseHelper.Database.SaveProduct(product);

            _productList = DatabaseHelper.Database.GetProducts();
        }

        private Boolean DeleteSelectedItems()
        {
            Boolean isSelected = false;
            for (int i = 0; i < mListView.Count; i++)
            {
                View rowView = mListView.GetChildAt(i);
                CheckBox cb = (CheckBox)rowView.FindViewById(Resource.Id.checkbox);
                if (cb.Checked)
                {
                    isSelected = true;
                    ProductAdapter.ViewHolder holder = (ProductAdapter.ViewHolder)rowView.Tag;
                    DatabaseHelper.Database.DeleteProduct(holder.ID);
                }
            }
            return isSelected;
        }
    }

    // Product Adapter
    public class ProductAdapter : BaseAdapter
    {
        private Context context;
        private int resource;
        private List<Product> productList;
        private ListViewMode mode;
        private LayoutInflater inflater;

        public ProductAdapter(Context context, int resource, List<Product> productList, ListViewMode mode)
        {
            this.context = context;
            this.resource = resource;
            this.productList = productList;
            this.mode = mode;
            this.inflater = (LayoutInflater)context.GetSystemService(Context.LayoutInflaterService);
        }

        public override int Count
        {
            get { return productList.Count; }
        }

        public override Java.Lang.Object GetItem (int position)
        {
            return null;
            //return productList[position];
        }

        public override long GetItemId (int position)
        {
            return productList[position].ProductId;
        }

        public override View GetView(int position, View convertView, ViewGroup parent)
        {
            ViewHolder holder;
            View row = convertView;
            if (row == null)
            {
                row = inflater.Inflate(this.resource, parent, false);
                holder = new ViewHolder();
                holder.ProductName = (TextView)row.FindViewById(Resource.Id.productname);
                holder.Price = (TextView)row.FindViewById(Resource.Id.price);
                holder.Image = (ImageView)row.FindViewById(Resource.Id.image);
                holder.Selected = (CheckBox)row.FindViewById(Resource.Id.checkbox);
                row.Tag = holder;
            }
            else
            {
                holder = (ViewHolder)row.Tag;
            }

            Product product = productList[position];
            holder.ID = product.ProductId;
            holder.ProductName.Text = product.ProductName;
            holder.Price.Text = "$" + Convert.ToString(product.Price);
            holder.Image.SetImageBitmap(ImageHelper.BytesToBitmap(product.Image));
            if (mode == ListViewMode.Display)
            {
                holder.Selected.Visibility = ViewStates.Gone;
            }
            else
            {
                holder.Selected.Visibility = ViewStates.Visible;
            }
            row.SetOnClickListener(new ProductItemOnClickListener(context, product.ProductId));
            return row;
        }

        public class ViewHolder : Java.Lang.Object 
        {
            public int ID { get; set; }
            public TextView ProductName { get; set; }
            public TextView Price { get; set; }
            public ImageView Image { get; set; }
            public CheckBox Selected { get; set; }
        }
    }

    class ProductItemOnClickListener : Java.Lang.Object, IOnClickListener
    {
        private Context context;
        private int id;

        public ProductItemOnClickListener(Context context, int id)
        {
            this.context = context;
            this.id = id;
        }

        public void OnClick(View view)
        {
            var intent = new Intent(this.context, typeof(ProductDetailActivity));
            intent.PutExtra(GameStoreConstants.ParamAction, GameStoreConstants.ParamActionEdit);
            intent.PutExtra(GameStoreConstants.ParamId, id);
            this.context.StartActivity(intent);
        }
    }
}
